[chiageng] iP

chiageng opened the pull request:
## Pull Request: Enhancements to Duke
[Fork Source](https://github.com/nus-cs2103-AY2324S2/ip)

>“Your mind is for having ideas, not holding them.” 
### Description

🚀 **Extended Features: AI Chatbot for Task Management**

This pull request introduces significant enhancements to Duke, transforming it into a powerful AI chatbot for managing to-do lists, events, and deadline tasks.

### Changes Made

🔄 **OOP Structure and Exception Handling**

- Implemented a more robust Object-Oriented Programming (OOP) structure for improved code modularity.
- Added exception handling to notify users of invalid or incorrect commands.

📁 **Persistent Storage**

- Tasks are now stored in the hard disk, ensuring data persistence for future access.

🧪 **Junit Testing**

- Introduced comprehensive Junit testing to validate and verify every feature of the application.

### New Features

📝 **Task Types**

Duke can now create three different types of tasks:

1. **To Do Task**
2. **Deadline Task**
3. **Event Task**

📅 **Date and Time Display**

When displaying date and time-sensitive tasks, Duke provides detailed information for better task management.
![image](https://github.com/nus-cs2103-AY2324S2/ip/assets/91059092/a2fad467-33a6-498e-b644-ad7a4f5f7933)



### Checklist

- [x] All tests pass
- [x] Code follows project coding standards
- [x] Documentation is updated
- [x] Changes have been tested on relevant browsers or platforms

### Code Example

```java
public static void main(String[] args) {
    new Duke().startChat();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

chiageng committed:
Give chatbot a new name and implement an initial skeletal version. Ask git to ignore *.class file in future.

chiageng committed:
Improve the skeletal version of Duke so that it exit when user type command bye

chiageng committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

chiageng committed:
Add a Task class. Add the ability to mark task as done and not done.

chiageng committed:
Add support for tracking three types of tasks

chiageng committed:
Automated Text UI Testing

chiageng committed:
Create DukeException class. Handle errors. Update automated texting and works fine now.

chiageng committed:
Add support for deleting tasks from the list.

chiageng committed:
Change constants to ENUM type

chiageng committed:
Format coding styles and javadocs

chiageng committed:
Store list of task into .txt. Load .txt file when chatbot start.

chiageng committed:
Merge branch 'branch-Level-7'

chiageng committed:
Use DateTime for deadline and event period.

chiageng committed:
Merge branch 'branch-Level-8'

chiageng committed:
Update the code to be more OOP.

chiageng committed:
Merge branch 'branch-A-MoreOOP'

chiageng committed:
Divide classes into packages.

chiageng committed:
Merge branch 'branch-A-Packages'

chiageng committed:
Merge branch 'add-gradle-support'

chiageng committed:
Run Duke using Gradle.

chiageng committed:
Merge branch 'branch-A-Gradle'

chiageng committed:
Create Junit Test on Tasks, Tasklist, Deadline, ToDo, Event using Gradle.

chiageng committed:
Merge branch 'branch-A-Junit'

chiageng committed:
Update build.gradle file.

chiageng committed:
Update javadocs of every public methods.

chiageng committed:
Merge tag 'A-JavaDoc'

chiageng committed:
Update code to follow coding standard.

chiageng committed:
Merge branch 'branch-A-CodingStandard'

chiageng committed:
Create new function find.

chiageng committed:
Merge branch 'branch-Level-9'

chiageng committed:
Add checkstyle.xml and suppressions.xml

chiageng committed:
Merge branch 'branch-A-CheckStyle'

Joseph31416 reviewed (commented):
Overall, very well done, excellent adherence to coding standards. 
I like how the commands area split into different classes and they inherit from the `Command` class. 
I like the fact that JavaDocs are provided for each class and method.

chiageng committed:
Add GUI to Duke

chiageng committed:
Add functions to GUI

chiageng committed:
Update fxml design

chiageng committed:
Merge branch 'branch-Level-10'

kjw142857 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * AddCommand Class
+ */
+public class AddCommand extends Command {
```
Here, `AddCommand` sounds more like an action (adding a command) rather than a class name (noun) to me. Perhaps you might want to rename this class to `TaskAdditionCommand`?

kjw142857 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * FindCommand class
+ */
+public class FindCommand extends Command {
```
Similarly, here `FindCommand` sounds a little like the action of finding a command to me! Maybe you could rename it as `TaskFinderCommand`?

kjw142857 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public String getExecuteMessage(TaskList tasks, Ui ui, Storage storage) {
```
Here, it is not entirely clear to me at first glance if the method `getExecuteMessage` is getting the message or executing it. Would it be possible to rename it is `getExecutionMessage`?

kjw142857 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,336 @@
+    /**
+     * Lists out all the task from the storage.
+     */
+    public void listTask() {
```
Here, as you are listing many tasks, I feel that I would like a plural form better. Would you be open to renaming this method as `listTasks`?

kjw142857 reviewed (commented):
Overall, most of your method and class names are appropriate and sensible choices! A few of them have some ambiguity (you can see my comments below), especially those related to class names inheriting the `Command` class - as writing a verb in front of the command noun, such as `AddCommand`, may lead to confusion with the action of adding a command. Otherwise, LGTM!

chiageng committed:
Add assertion to validate input parameter

The code currently lacks validation for input passed
to some functions such as getResponse which may lead to
unexpected behaviour.

Add assertions at the beginning of function receiving user
input, ensuring expected criteria before proceeding with
the execution.

Using assertions in this manner helps catch potential issues
early in the development process, making it easier to identify
and fix bugs.

chiageng committed:
Merge pull request #1 from chiageng/branch-A-Assertions

Add assertion to validate input parameter

chiageng committed:
Refractor code to comply with coding standards

Some naming functions are hard to understand.

Refractor the function name such as getExecution to
getExecutionMessage and getResponse to getResponseMessage.

By aligning with coding standards, the naming is now more
intuitive and easier to understand.

chiageng committed:
Refractor code to comply with coding standards

The previous class naming is confusing, such as AddCommand
leading to misunderstanding of adding a new command.

Refractor Command class name such as AddCommand to
TaskAddingCommand which is more intuitive.

By aligning the code with coding standards, it improve
readability and easier to understand.

chiageng committed:
Reformat code to comply with coding standard

chiageng committed:
Merge pull request #2 from chiageng/branch-A-CodeQuality

Refractor code to comply with coding standard and coding quality

chiageng committed:
Implement Undo Command

chiageng committed:
Merge branch 'branch-BCD-Extension'

chiageng committed:
Allow dialog box to display full messages

chiageng committed:
Change duke font to itallic

chiageng committed:
Display error messages with red background

chiageng committed:
Merge branch 'branch-A-BetterGui'

chiageng committed:
Give new name as ChatPal and friendly personality

chiageng committed:
Merge branch 'branch-A-Personality'

chiageng committed:
Add exception, event start time cannot be later than end time

chiageng committed:
Merge pull request #4 from chiageng/branch-A-MoreErrorHandling

Add exception, event start time cannot be later than end time

chiageng committed:
Merge branch 'master' of https://github.com/chiageng/ip

chiageng committed:
Add more Junit Tests

chiageng committed:
Merge pull request #5 from chiageng/branch-A-MoreTesting

Add more Junit Tests

chiageng committed:
Merge branch 'master' of https://github.com/chiageng/ip

chiageng committed:
Merge branch 'branch-AiAssisted'

chiageng committed:
Update README.md

chiageng committed:
Create new jar file
